nixos/modules/services/web-servers/nginx/cgit.nix

```
@@ -0,0 +1,130 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  globalConfig = config;
+  settingsFormat = {
+    type = with lib.types; let
+      value = oneOf [ int str ] // {
+        description = "INI-like atom (int or string)";
+      };
+      values = coercedTo value lib.singleton (listOf value) // {
+        description = value.description + " or a list of them for duplicate keys";
+      };
+    in
+    attrsOf (values);
+    generate = name: values:
+      pkgs.writeText name (lib.generators.toKeyValue { listsAsDuplicateKeys = true; } values);
+  };
+in
+{
+  options.services.nginx.virtualHosts = mkOption {
+    type = types.attrsOf (types.submodule ({ config, ... }:
+      let
+        cfg = config.cgit;
+
+        # These are the global options for this submodule, but for nicer UX they
+        # are inlined into the freeform settings.  Hence they MUST NOT INTERSECT
+        # with any settings from cgitrc!
+        options = {
+          enable = mkEnableOption "cgit";
+
+          location = mkOption {
+            default = "/";
+            type = types.str;
+            description = ''
+              Location to serve cgit on.
+            '';
+          };
+        };
+
+        # Remove the global options for serialization into cgitrc
+        settings = removeAttrs cfg (attrNames options);
+      in
+      {
+        options.cgit = mkOption {
+          type = types.submodule {
+            freeformType = settingsFormat.type;
+            inherit options;
+            config = {
+              css = mkDefault "/cgit.css";
+              logo = mkDefault "/cgit.png";
+              favicon = mkDefault "/favicon.ico";
+            };
+          };
+          default = { };
+          example = literalExample ''
+            {
+              enable = true;
+              virtual-root = "/";
+              source-filter = "''${pkgs.cgit}/lib/cgit/filters/syntax-highlighting.py";
+              about-filter = "''${pkgs.cgit}/lib/cgit/filters/about-formatting.sh";
+              cache-size = 1000;
+              scan-path = "/srv/git";
+              include = [
+                (builtins.toFile "cgitrc-extra-1" '''
+                  # Anything that has to be in a particular order
+                ''')
+                (builtins.toFile "cgitrc-extra-2" '''
+                  # Anything that has to be in a particular order
+                ''')
+              ];
+            }
+          '';
+          description = ''
+            Verbatim contents of the cgit runtime configuration file. Documentation
+            (with cgitrc example file) is available in "man cgitrc". Or online:
+            http://git.zx2c4.com/cgit/tree/cgitrc.5.txt
+          '';
+        };
+
+        config = mkIf cfg.enable {
+
+          locations."${cfg.location}" = {
+            root = "${pkgs.cgit}/cgit/";
+            tryFiles = "$uri @cgit";
+          };
+          locations."~* ^/(cgit.(css|png)|favicon.ico|robots.txt)" = {
```

I think this actually should be
```suggestion
          locations."~ ^${cfg.location}/(cgit.(css|png)|favicon.ico|robots.txt)" = {
```
but I'm not 100% sure since I'm not too familiar with nginx's location matching.  From what I know `~*` would be a case-insensitive regex match, which seems wrong and `^/(...)` would only match correctly if `location = "/"`.  But again, I could be wrong.

---

You are right,
this can be done with :
`locations."~ ^${cfg.location}/(cgit.(css|png)|favicon.ico|robots.txt)$" = {
    alias = "${pkgs.cgit}/cgit/$1";
}; `
We also need to do a test and change default cgitrc values, but i lack skill with nix.
I'm trying to do this:
` css = mkDefault "${cfg.options.location}/cgit.css";`
But i get "infinite recursion encountered" error.
i need to learn nix in the meantime, help would be welcome.

---

I think i found :-), I will be able to test

---

maybe the problem is not that trivial in fact. anyway it's just default options and this shouldn't cause  backward compatibility problem in case of improvement.